compiler/src/iree/compiler/Codegen/Common/GPU/AMDGPULowerCoalescedDMAToGatherLDS.cpp

```diff
@@ -0,0 +1,246 @@
+        targetDmaSizes(targetDmaSizes) {}
+
+  LogicalResult matchAndRewrite(IREE::GPU::CoalescedGatherDMAOp dmaOp,
+                                PatternRewriter &rewriter) const override {
+    LDBG() << "Processing CoalescedGatherDMAOp: " << dmaOp;
+
+    auto forallOp = dmaOp->getParentOfType<scf::ForallOp>();
+    if (!forallOp) {
+      return rewriter.notifyMatchFailure(
+          dmaOp, "coalesced_gather_dma not inside scf.forall");
+    }
+
+    if (failed(verifyThreadMapping(forallOp))) {
+      return rewriter.notifyMatchFailure(
+          dmaOp, "forall does not have proper thread mapping");
+    }
+
+    if (failed(verifyMemoryLayoutContiguous(dmaOp, rewriter))) {
+      return failure();
+    }
+
+    Value source = dmaOp.getSource();
+    Value dest = dmaOp.getInit();
+    OperandRange indicesRange = dmaOp.getIndices();
+
+    auto sourceType = cast<MemRefType>(source.getType());
+    auto destType = cast<MemRefType>(dest.getType());
+
+    if (!indicesRange.empty()) {
+      Value indices = indicesRange.front();
+      auto indicesType = cast<MemRefType>(indices.getType());
+      ArrayRef<int64_t> indicesShape = indicesType.getShape();
+      ArrayRef<int64_t> destShape = destType.getShape();
+
+      // Verify that indices dimensions match dest dimensions.
+      if (indicesShape != destShape) {
+        return rewriter.notifyMatchFailure(
+            dmaOp, "indices shape does not match dest shape");
+      }
+    }
+
+    Type elementType = sourceType.getElementType();
+    int64_t elementBits = sourceType.getElementTypeBitWidth();
+    LDBG() << "  Element type: " << elementType;
+    LDBG() << "  Element bits: " << elementBits;
+
+    int64_t innermostDimSize = sourceType.getShape().back();
+    int64_t transferSizeBits = innermostDimSize * elementBits;
+    LDBG() << "  Source innermost dimension size: " << innermostDimSize;
+    LDBG() << "  Transfer size in bits: " << transferSizeBits;
+
+    std::optional<int64_t> subgroupSize =
+        getSubgroupSize(dmaOp->getParentOfType<FunctionOpInterface>());
+    if (!subgroupSize.has_value()) {
+      return rewriter.notifyMatchFailure(
+          dmaOp, "unable to determine subgroup size from forall");
+    }
+    LDBG() << "Subgroup size: " << *subgroupSize;
+
+    // Check that transfer size matches one of the target DMA sizes.
+    int64_t transferSizePerLane = transferSizeBits / *subgroupSize;
+    LDBG() << "Transfer size per lane: " << transferSizePerLane << " bits";
+
+    if (!targetDmaSizes.empty() &&
+        !llvm::is_contained(targetDmaSizes, transferSizePerLane)) {
+      return rewriter.notifyMatchFailure(
+          dmaOp, "transfer size does not match any target DMA size");
+    }
+    LDBG() << "Transfer size matches target DMA sizes";
+
+    // TODO: Handle indices properly - for now we skip the explicit indices
+    // check The lane parameter is always present but we handle it differently.
```

Comment:
I don't follow this comment.

Also, if non-copy gathers aren't supported, we should error out.